druntime/src/core/internal/array/capacity.d

```diff
@@ -9,6 +9,181 @@
 */
 module core.internal.array.capacity;
 
+import core.attribute : weak;
+
+// for now, all GC array functions are not exposed via core.memory.
+extern (C)
+{
+    size_t gc_reserveArrayCapacity(void[] slice, size_t request, bool atomic) nothrow pure;
+    bool gc_shrinkArrayUsed(void[] slice, size_t existingUsed, bool atomic) nothrow pure;
+}
+
+
+private uint __typeAttrs(T)(void *copyAttrsFrom = null)
+{
+    import core.internal.traits : hasElaborateDestructor, hasIndirections;
+    import core.memory : GC;
+
+    alias BlkAttr = GC.BlkAttr;
+
+    if (copyAttrsFrom)
+    {
+        // try to copy attrs from the given block
+        auto info = GC.query(copyAttrsFrom);
+        if (info.base)
+            return info.attr;
+    }
+
+    uint attrs = 0;
+    static if (hasIndirections!T)
+        attrs |= BlkAttr.NO_SCAN;
+
+    static if (hasElaborateDestructor!T)
+        attrs |= BlkAttr.FINALIZE;
+
+    return attrs;
+}
+
+private void __doPostblit(T)(T[] arr)
+{
+    // infer static postblit type, run postblit if any
+    static if (__traits(hasPostblit, T))
+    {
+        static if (__traits(isStaticArray, T) && is(T : E[], E))
+            __doPostblit(cast(E[]) arr);
+        else static if (!is(typeof(arr[0].__xpostblit())) && is(immutable T == immutable U, U))
```

Comment:
Isn't `!__traits(compiles, arr[0].__xpostblit())` what you're looking for instead of `!is(typeof(...))`?

Second: will this ever be used for `immutable` types? From a purely theoretical standpoint it's wrong, but so is `_d_arraysetcapacityPureNothrow`, yet here we are. My point is let's break the type system as little as we can.